examples/dampened_oscillator/dampened_oscillator_declarative.py

```diff
@@ -0,0 +1,133 @@
+from numerous.multiphysics.equation_decorators import Equation
+from numerous.multiphysics.equation_base import EquationBase
+from numerous.engine.system.item import Item
+from numerous.engine.system import Subsystem, ConnectorTwoWay, ItemsStructure
+from numerous.declarative.specification import ScopeSpec, ItemsSpec, Module, EquationSpec
+from numerous.declarative.variables import Parameter, Constant, State
+from numerous.declarative.bus import Connector, create_connections, get_value_for, set_value_from
+
+import numpy as np
+
+class DampenedOscillator(Module):
+    """
+        Equation and item modelling a spring and dampener
+    """
+    tag = "dampened_oscillator"
+
+    class Mechanics(ScopeSpec):
+        # Define variables
+        k = Constant(1) # spring constant
+        c = Constant(1) # coefficient of friction
+        a = Parameter(0) # acceleration
+        x, x_dot = State(0) # distance
+        v, v_dot = State(0) # velocity
+
+    mechanics = Mechanics()
+
+    coupling = Connector(
+        x = set_value_from(mechanics.x),
+        F = get_value_for(mechanics.v_dot)
+    )
+
+    @EquationSpec(mechanics)
+    def eval(self, scope: Mechanics):
+        scope.a = -scope.k * scope.x - scope.c * scope.v
+        scope.v_dot = scope.a
+        scope.x_dot = scope.v
+
+
+class SpringCoupling(Module):
+
+    tag = "springcoup"
+
+    class Mechanics(ScopeSpec):
+        k = Parameter(1)
+        c = Parameter(1)
+        F1 = Parameter(0)
+        F2 = Parameter(0)
+        x1 = Parameter(0)
+        x2 = Parameter(0)
+
+    mechanics = Mechanics()
+
+    class Items(ItemsSpec):
+        side1: DampenedOscillator
+        side2: DampenedOscillator
+
+    side1 = Connector(
+        F = set_value_from(mechanics.F1),
+        x = get_value_for(mechanics.x1)
+    )
+
+    side2 = Connector(
+        F=set_value_from(mechanics.F2),
+        x=get_value_for(mechanics.x2)
+    )
+
+    items = Items()
+
+    def __init__(self, tag="springcoup", k=1.0):
+        super().__init__(tag)
+
+        self.mechanics.set_values(k=k)
+
+    @EquationSpec(mechanics)
+    def eval(self, scope: Mechanics):
+        scope.c = scope.k
+
+        dx = scope.x1 - scope.x2
+        F = np.abs(dx) * scope.c
+
+        scope.F1 = -F if scope.x1 > scope.x2 else F  # [kg/s]
+
+        scope.F2 = -scope.F1
+
+
+class OscillatorSystem(Module):
+
+    class Items(ItemsSpec):
+        # The oscillator system will have two oscillators interacting via the coupling
+        oscillator1: DampenedOscillator
+        oscillator2: DampenedOscillator
+        coupling: SpringCoupling
+
+    items = Items()
```

Comment:
Is it by design that you are not using dependency injection? But statically defining the items here?

Reply:
I don't know exactly what you mean regarding dependecy injection. But yes the concept is to declare the items which will later be added to the model. This allows the IDE to perform autocompletion and type checking when implementing the model in the __init__ function. It is still possible to dynamically create items inside eg. the __init__ function or use dependency injection to supply the items to the __init__ function.